Sources/AsyncHTTPClient/ConnectionPool/HTTP1ProxyConnectHandler.swift

```diff
@@ -0,0 +1,143 @@
+
+        self.sendConnect(context: context)
+    }
+
+    func handlerRemoved(context: ChannelHandlerContext) {
+        switch self.state {
+        case .failed, .completed:
+            break
+        case .initialized, .connectSent, .headReceived:
+            preconditionFailure("Removing the handler, while connecting seems wrong")
+        }
+    }
+
+    func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        preconditionFailure("We don't support outgoing traffic during HTTP Proxy update.")
+    }
+
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        switch self.unwrapInboundIn(data) {
+        case .head(let head):
+            guard case .connectSent(let promise) = self.state else {
+                preconditionFailure("HTTPDecoder should throw an error, if we have not send a request")
+            }
+
+            switch head.status.code {
+            case 200..<300:
+                // Any 2xx (Successful) response indicates that the sender (and all
+                // inbound proxies) will switch to tunnel mode immediately after the
+                // blank line that concludes the successful response's header section
+                self.state = .headReceived(promise)
+            case 407:
+                let error = HTTPClientError.proxyAuthenticationRequired
+                self.state = .failed(error)
+                context.close(promise: nil)
+                promise.fail(error)
+            default:
+                // Any response other than a successful response
+                // indicates that the tunnel has not yet been formed and that the
+                // connection remains governed by HTTP.
+                let error = HTTPClientError.invalidProxyResponse
+                self.state = .failed(error)
+                context.close(promise: nil)
+                promise.fail(error)
+            }
+        case .body:
+            switch self.state {
+            case .headReceived(let promise):
+                // we don't expect a body
+                let error = HTTPClientError.invalidProxyResponse
+                self.state = .failed(error)
+                context.close(promise: nil)
+                promise.fail(error)
+            case .failed:
+                // ran into an error before... ignore this one
+                break
+            case .completed, .connectSent, .initialized:
+                preconditionFailure("Invalid state")
+            }
+
+        case .end:
+            switch self.state {
+            case .headReceived(let promise):
+                self.state = .completed
+                promise.succeed(())
+            case .failed:
+                // ran into an error before... ignore this one
+                break
+            case .initialized, .connectSent, .completed:
+                preconditionFailure("Invalid state")
+            }
+        }
+    }
```

Comment:
Again, personal opinion, this code is perhaps a little more difficult to read than it needs to be. Could you refactor out into separate functions to handle each http part?